src/angular/11-declarative-state/restaurant.component.ts

```diff
@@ -1,85 +1,157 @@
+        toData,
+        startWith({ isPending: true, value: [] }),
+        shareReplay({ bufferSize: 1, refCount: true })
+      );
+
+    this.enableStateSelect$ = this.states$.pipe(
+      takeUntil(this.onDestroy$),
+      tap((states) => {
+        if (states.value.length > 0) {
+          this.form.get('state')!.enable();
+        }
+      })
     );
 
-    this.cities = this.selectedState.pipe(
-        flatMap((state) => {
-          if (state) {
-            return this.restaurantService.getCities(state).pipe(
-              toData,
-              startWith({ isPending: true, value: [] })
-            )
-          } else {
-            return of({ isPending: false, value: [] });
-          }
-        }),
-        shareReplay(1),
+    this.cities$ = this.selectedState$.pipe(
+      mergeMap((state) => {
+        if (state) {
+          return this.restaurantService
+            .getCities(state)
+            .pipe(toData, startWith({ isPending: true, value: [] }));
+        } else {
+          return of({ isPending: false, value: [] });
+        }
+      }),
+      shareReplay({ bufferSize: 1, refCount: true })
     );
-    this.citySelectDisabled = this.cities.pipe(
-      map(cities => cities.value.length === 0)
+
+    this.toggleCitySelect$ = this.cities$.pipe(
+      takeUntil(this.onDestroy$),
+      tap((cities) => {
+        if (cities.value.length === 0) {
+          this.form.get('city')!.disable({
+            onlySelf: true,
+            emitEvent: false,
+          });
+        } else {
+          this.form.get('city')!.enable({
+            onlySelf: true,
+            emitEvent: false,
+          });
+        }
+      })
     );
-    this.displayedCity = merge(
-      this.selectedState.pipe(map(() => '')),
-      this.selectedCity
+
+    this.clearCityWhenStateChanges$ = this.selectedState$.pipe(
+      takeUntil(this.onDestroy$),
+      pairwise(),
+      tap(([previous, current]) => {
+        if (current && current !== previous) {
+          this.form.get('city')!.patchValue('');
+        }
+      })
     );
 
-    this.restaurants = combineLatest(this.displayedCity, this.selectedState).pipe(
-      flatMap(([city, state]) => {
+    this.restaurants$ = combineLatest([
+      this.selectedCity$,
+      this.selectedState$,
+    ]).pipe(
+      mergeMap(([city, state]) => {
         if (city && state) {
-          return this.restaurantService.getRestaurants(state, city).pipe(
-            toData,
-            startWith({ isPending: true, value: [] })
-          )
+          return this.restaurantService
+            .getRestaurants(state, city)
+            .pipe(toData, startWith({ isPending: true, value: [] }));
         } else {
           return of({ isPending: false, value: [] });
         }
-      }),
-      shareReplay(1),
+      })
     );
   }
+
+  ngOnInit(): void {
+    this.enableStateSelect$.subscribe();
+    this.toggleCitySelect$.subscribe();
+    this.clearCityWhenStateChanges$.subscribe();
+  }
```

Comment:
not super in love with this implementation - I feel like it's easy to miss the subscribes here. also - some of these can just be consts vs. members on the component if they're only used here.

Reply:
for this reason alone I would recommend to always subscribe OnInit instead of constructor

Reply:
then constructor is just for setting members and not to run logic. run logic using component lifecycle methods

Reply:
ok, i can get behind that